contributors/design-proposals/auth/bound-service-account-tokens.md

```diff
@@ -0,0 +1,179 @@
+to request JSON Web Tokens that are audience, time and eventually key bound.
+
+# Background
+
+Kubernetes already provisions JWTs to workloads. This functionality is on by
+default and thus widely deployed. The current workload JWT system has serious
+issues:
+
+1.  Security: JWTs are not audience bound. Any recipient of a JWT can masquerade
+    as the presenter to anyone else.
+1.  Security: The current model of storing the service account token in a Secret
+    and delivering it to nodes results in a broad attack surface for the
+    Kubernetes control plane when powerful components are run - giving a service
+    account a permission means that any component that can see that service
+    account's secrets is at least as powerful as the component.
+1.  Security: JWTs are not time bound. A JWT compromised via 1 or 2, is valid
+    for as long as the service account exists. This may be mitigated with
+    service account signing key rotation but is not supported by client-go and
+    not automated by the control plane and thus is not widely deployed.
+1.  Scalability: JWTs require a Kubernetes secret per service account.
+
+# Proposal
+
+Infrastructure to support on demand token requests will be implemented in the
+core apiserver. Once this API exists, a client of the apiserver will request an
+attenuated token for it's own use. The API will enforce required attenuations,
+e.g. audience and time binding.
+
+## Token attenuations
+
+### Audience binding
+
+Tokens issued from this API will be audience bound. Audience of requested tokens
+will be bound by the `aud` claim. The `aud` claim is an array of strings
+(usually URLs) that correspond to the intended audience of the token. A
+recipient of a token is responsible for verifying that it identifies as one of
+the values in the audience claim, and should otherwise reject the token. The
+TokenReview API will support this validation.
+
+### Time binding
+
+Tokens issued from this API may be time bound. Time validity of these tokens
+will be claimed in the following fields:
+
+* `exp`: expiration time
+* `nbf`: not before
+* `iat`: issued at
+
+A recipient of a token should verify that the token is valid at the time that
+the token is presented, and should otherwise reject the token. The TokenReview
+API will support this validation.
+
+### Key binding
+
+Tokens issued from this API will OPTIONALLY (per the behest of the requestor)
+be key bound. The `cnf` claim will carry information that a recipient of a
+token might need to validate that the token presenter possesses a specific key.
+A recipient may require accepted tokens are key bound.
+
+## API Changes
+
+### Add `tokenrequests.authentication.k8s.io`
+
+We will add an imperative API (a la TokenReview) to the
+`authentication.k8s.io` API group:
+
+```golang
+type TokenRequest struct {
+  Spec   TokenRequestSpec
+  Status TokenRequestStatus
+}
+
+type TokenRequestSpec struct {
```

Comment:
all the use cases I can think of for token requests benefit from including kind+name+uid of at least one object as a claim in the token:
* a token requested by node `n` for service account `sa` should be bound to a pod with `serviceAccountName=sa` and `nodeName=n` (and stop working against the apiserver when that pod goes away)
* a non-expiring token should be bound to some object in the namespace that indicates the existence of the token, and can be deleted in order to revoke the token (just like a service account token secret behaves today)

What if the TokenRequestSpec included an object reference (or list of object references) that the token should be bound to? That would let the NodeRestriction plugin limit nodes to requesting tokens bound to pods scheduled to themselves and referencing the correct serviceaccount.

Reply:
Should this be an ObjectReference or LocalObjectReference (would have to add a few fields to LocalObjectReference) or should I create a new type for this?

Reply:
Added a section on object binding.